File: PipeliningWithGenerators/Pipeline.cs
// Pipelining with generators.
using System;


class Pipeline 
{
    public static IEnumerable<int> Generate() 
    {
        for (int i = 0; i < 10; ++i)
            yield return i;
    }

    public static IEnumerable<int> Square(IEnumerable<int> seq) 
    {
        foreach (var k in seq)
            yield return k * k;
    }

    public static IEnumerable<int> SquareLINQ(IEnumerable<int> seq)
        => from k in seq select k * k;

    static int Sum(IEnumerable<int> seq) 
    {
        int s = 0;

        foreach (var k in seq)
            s += k;
        
        return s;
    }

    static void Main() 
    {
        Sum(Square(Generate())).Dump();
        Sum(SquareLINQ(Generate())).Dump();
    }
}

File: PipeliningWithGenerators/pipeline.py
# Pipelining with generators.

import sys


def generate():
    for k in range(10):
        print('G', end='')
        yield k


def square(seq):
    for k in seq:
        print('Q', end='')
        yield k * k


def summ(seq):
    s = 0
    for k in seq:
        print('S', end='')
        s += k
    return s


def main():
    print(summ(square(generate())))
    return 0


"""
The logging will not output GGGGGGGGGGQQQQQQQQQQSSSSSSSSSS285
but rather GQSGQSGQSGQSGQSGQSGQSGQSGQSGQS285
indicating that the coroutines 'generate' and 'square' are executed
interleavingly with the function 'summ' that invokes them. 
"""
if __name__ == "__main__":
    sys.exit(main());
